### PR TITLE
[Merged by Bors] - TY-2360 Build android libs on bors CI

### PR DESCRIPTION
**Pull request by x3ro**
[TY-2360]

[TY-2360]: https://xainag.atlassian.net/browse/TY-2360?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -13,19 +13,81 @@ permissions:
   contents: read
```
Comment (acrrd) [inline .github/workflows/ci.yml:13]: This is not where we want this but we want it in `ci_bors.yml`.

File: discovery_engine_core/rust-toolchain.toml [hunk 2]
```diff
@@ -1,3 +1,4 @@
 [toolchain]
 channel = "1.55.0"
 profile = "default"
+targets = ["aarch64-apple-ios", "x86_64-apple-ios"]
```
Comment (acrrd) [inline discovery_engine_core/rust-toolchain.toml:4]: This will force to install the targets in evrey job but we want to have them only in few of them so we can have an action or step to do it.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (x3ro) [reply .github/workflows/ci.yml:13]: This is only so I can do test runs while developing.

**Comment by x3ro** [comment]
bors try

**Comment by x3ro** [comment]
bors try

**Comment by x3ro** [comment]
bors try-

**Comment by x3ro** [comment]
bors try

**Comment by x3ro** [comment]
bors try

File: .env [hunk 3]
```diff
@@ -9,8 +9,13 @@ FLUTTER_WORKSPACE=discovery_engine_flutter
 DART_VERSION=2.14.4
 FLUTTER_VERSION=2.5.3
 
-ANDROID_TARGETS="arm64-v8a x86_64 x86"
```
Comment (x3ro) [inline .env:12]: I've removed this target to improve build times for now, as I don't really know if we need a non 64-bit Intel target, considering that we're e.g. not supporting armv7. Let me know if that is not a good idea.

File: .github/actions/setup-job/action.yml [hunk 4]
```diff
@@ -50,6 +54,18 @@ runs:
         flutter-version: ${{ env.FLUTTER_VERSION }}
         cache: true
 
+    - name: Install Android NDK
+      if: inputs.android != 'false'
+      uses: nttld/setup-ndk@4a541f506ff27d545f89e860b7cdc86efc3ca496 # HEAD 2022-02-15
+      with:
+        ndk-version: ${{ env.ANDROID_NDK_VERSION }}
```
Comment (Robert-Steiner) [inline .github/actions/setup-job/action.yml:61]: We shouldn't need that. The Ubuntu runner should have ndk installed [ref](https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-Readme.md#environment-variables-3)

File: .github/actions/setup-job/action.yml [hunk 5]
```diff
@@ -50,6 +54,18 @@ runs:
...
+
+    - name: Install cargo-ndk
+      shell: bash
+      if: inputs.android != 'false'
+      run: |
+        cargo install --git https://github.com/xaynetwork/cargo-ndk.git cargo-ndk
```
Comment (Robert-Steiner) [inline .github/actions/setup-job/action.yml:67]:
```suggestion
      run: cargo install --git https://github.com/xaynetwork/cargo-ndk.git cargo-ndk
```

File: .github/workflows/ci_bors.yml [hunk 6]
```diff
@@ -19,15 +19,22 @@ jobs:
...
     runs-on: ubuntu-20.04
     timeout-minutes: 5
     if: always()
     steps:
       - name: CI OK Testing
         run: |
-          if [[ ${{ needs.dev-ci.outputs.wf-result }} == "success" && "${{ needs.build-ios-libs.result }}" == "success" ]]; then
+          if [[ \
+            "${{ needs.dev-ci.outputs.wf-result }}" == "success" && \
+            "${{ needs.build-ios-libs.result }}" == "success" &&
```
Comment (Robert-Steiner) [inline .github/workflows/ci_bors.yml:35]: do we also need a `\` here or does it work without?
```suggestion
            "${{ needs.build-ios-libs.result }}" == "success" && \
```

File: .github/workflows/ci_build_android.yml [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        description: The base name of artifact directory (without the target suffix)
+        value: ${{ jobs.build-android-libs.outputs.artifact-dir-base }}
+
+jobs:
+  build-android-libs:
+    runs-on: ubuntu-20.04
+    timeout-minutes: 35
+    strategy:
+      matrix:
+        target: ["arm64-v8a", "x86_64"]
```
Comment (Robert-Steiner) [inline .github/workflows/ci_build_android.yml:16]: In cargo ndk we can also use the [rust triple](https://github.com/bbqsrc/cargo-ndk/blob/f91b56d4114dd18fa7d2fa3b484fb214288e6450/src/meta.rs#L88). 
```suggestion
        target: ["aarch64-linux-android", "x86_64-linux-android"]
```

File: .github/workflows/ci_build_android.yml [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    steps:
+      - uses: actions/checkout@ec3a7ce113134d7a93b817d10a8272cb61118579 # v2.4.0
+
+      - name: Setup the rust CI.
+        uses: ./.github/actions/setup-job
+        with:
+          rust: true
+          dart: true
+          android: true
+          rust-target: "aarch64-linux-android x86_64-linux-android"
```
Comment (Robert-Steiner) [inline .github/workflows/ci_build_android.yml:30]: with the rust triple that can be then written as
```suggestion
          rust-target: ${{ matrix.target }}
```

File: .github/workflows/ci_build_android.yml [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    timeout-minutes: 35
+    strategy:
+      matrix:
+        target: ["arm64-v8a", "x86_64"]
+    env:
+      ARTIFACT_DIR_BASE: ${{ github.job }}
+    outputs:
+      artifact-dir-base: ${{ env.ARTIFACT_DIR_BASE }}
+    steps:
+      - uses: actions/checkout@ec3a7ce113134d7a93b817d10a8272cb61118579 # v2.4.0
```
Comment (Robert-Steiner) [inline .github/workflows/ci_build_android.yml:22]: @acrrd not part of this pr but i think we can also move the checkout to `setup-job`

File: .github/workflows/ci_build_android.yml [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        with:
+          rust: true
+          dart: true
+          android: true
+          rust-target: "aarch64-linux-android x86_64-linux-android"
+          github_token: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Build Android library
+        run: |
+          export PATH="$(pwd)/cargo-installs/bin:${PATH}"
```
Comment (Robert-Steiner) [inline .github/workflows/ci_build_android.yml:35]: why do we need that?

File: .github/workflows/ci_build_android.yml [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+          just compile-android "${{ matrix.target }}"
+
+      - name: Prepare lib for upload
+        id: lib
+        run: |
+          FILE=${{ env.FLUTTER_WORKSPACE }}/android/src/main/jniLibs/${{ matrix.target }}/libxayn_discovery_engine_bindings.so
```
Comment (Robert-Steiner) [inline .github/workflows/ci_build_android.yml:41]: we can also use 
```suggestion
          FILE=${{ env.FLUTTER_WORKSPACE }}/android/src/main/jniLibs/${{ matrix.target }}/${{ env.IOS_LIB_BASE }}.so
```
but then we should better remame it to `LIB_BASE` or something like that

File: justfile [hunk 12]
```diff
@@ -219,7 +219,7 @@ _dry-run-release: clean deps dart-build
 dry-run-release:
      @CI=true {{just_executable()}} _dry-run-release
 
-compile-android target:
+compile-android target: dart-build
```
Comment (Robert-Steiner) [inline justfile:222]: That probably fails because we need to run cargo check before in order to run the build scripts. can you please try if this would also work
```suggestion
compile-android target: _codegen-order-workaround
```

File: .github/workflows/ci_build_android.yml [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+          rust: true
+          dart: true
+          android: true
+          rust-target: "aarch64-linux-android x86_64-linux-android"
+          github_token: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Build Android library
+        run: |
+          export PATH="$(pwd)/cargo-installs/bin:${PATH}"
+          just compile-android "${{ matrix.target }}"
```
Comment (Robert-Steiner) [inline .github/workflows/ci_build_android.yml:36]:
```suggestion
          just compile-android ${{ matrix.target }}
```

File: .github/workflows/ci_build_android.yml [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+    outputs:
+      artifact-dir-base: ${{ env.ARTIFACT_DIR_BASE }}
+    steps:
+      - uses: actions/checkout@ec3a7ce113134d7a93b817d10a8272cb61118579 # v2.4.0
+
+      - name: Setup the rust CI.
+        uses: ./.github/actions/setup-job
+        with:
+          rust: true
+          dart: true
```
Comment (Robert-Steiner) [inline .github/workflows/ci_build_android.yml:28]:
```suggestion
```

**Review by Robert-Steiner — COMMENTED** [review]
looks good! added some small suggestions

File: .github/workflows/ci_build_android.yml (hunk 7 above)
Comment (acrrd) [reply .github/workflows/ci_build_android.yml:16]: On the CI we want build for all the android targets: aarch64-linux-android, armv7-linux-androideabi, i686-linux-android, x86_64-linux-android

Comment (x3ro) [reply .github/workflows/ci_build_android.yml:16]: Mmh, I think I was told that we don't support armv7. I'm assuming that information is incorrect, then? Also what's the point of building Intel non-64-bit architectures?

Comment (Robert-Steiner) [reply .github/workflows/ci_build_android.yml:16]: for ios we don't support armv7 because 32Bit is no longer supported since ios 11 i think

Comment (acrrd) [reply .github/workflows/ci_build_android.yml:16]: Someone still have an emulator that wants the non 64 version.

File: .github/workflows/ci_build_android.yml (hunk 10 above)
Comment (x3ro) [reply .github/workflows/ci_build_android.yml:35]: Because otherwise it doesn't find `cargo-ndk`

File: .github/workflows/ci_build_android.yml (hunk 13 above)
Comment (x3ro) [reply .github/workflows/ci_build_android.yml:36]: I think the general advice that shell linters (e.g. shellcheck) give is to quote variables / dynamic values at all times, to avoid unpleasant surprises. While this may not be needed here, I mostly try to follow that pattern :)

**Comment by x3ro** [comment]
bors try

File: .github/workflows/ci_build_android.yml (hunk 10 above)
Comment (Robert-Steiner) [reply .github/workflows/ci_build_android.yml:39]: that is weird because we never had to do this before 
also cargo-ndk shound't be install in `cargo-installs/bin` but in `~/.cargo/bin/`
`cargo-installs/bin` that is only for `async-bindgen-gen-dart`
can you try a run without that line please

File: .github/workflows/ci_build_android.yml (hunk 13 above)
Comment (Robert-Steiner) [reply .github/workflows/ci_build_android.yml]: you normally need quotes if the variable can be empty, contain spaces or wildcards. but I'm ok if you leave the quotes here

**Comment by x3ro** [comment]
bors try

File: .github/workflows/ci_build_android.yml [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+          github_token: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Build Android library
+        run: |
+          just compile-android "${{ matrix.target.ndk }}"
+
+      - name: Prepare lib for upload
+        id: lib
+        run: |
+          FILE=${{ env.FLUTTER_WORKSPACE }}/android/src/main/jniLibs/${{ matrix.target.ndk }}/libxayn_discovery_engine_bindings.so
```
Comment (Robert-Steiner) [inline .github/workflows/ci_build_android.yml:44]: i thing i forgot. to preserve the folder that cargo ndk creates for us we should change this to 
```suggestion
          FILE=${{ env.FLUTTER_WORKSPACE }}/android/src/main/jniLibs
```
so it will upload 
```
<arch_folder>/lib.so
```
instead of 
```
lib.so
```

we can then also remove this step and put this line directly to the upload action path

File: .github/workflows/ci_build_android.yml (hunk 10 above)
Comment (x3ro) [reply .github/workflows/ci_build_android.yml]: Breaks [here](https://github.com/xaynetwork/xayn_discovery_engine/runs/5214694142?check_suite_focus=true#step:4:25)

File: .github/workflows/ci_build_android.yml (hunk 11 above)
Comment (x3ro) [reply .github/workflows/ci_build_android.yml]: Outdated now because we'll upload the entire folder.

**Comment by x3ro** [comment]
bors try

**Comment by x3ro** [comment]
bors try-

**Comment by x3ro** [comment]
bors try

File: .github/workflows/ci_build_android.yml (hunk 10 above)
Comment (Robert-Steiner) [reply .github/workflows/ci_build_android.yml:39]: ah i think it because of `CARGO_INSTALL_ROOT` in `.env`. that will install cargo-ndk in `cargo-installs/bin`. @acrrd, @rustonaut is `cargo-installs/bin` only there for `async-bindgen-gen-dart` or was there the plan to also use it also for cargo-ndk, cargo sort,... ?

File: .github/workflows/ci_build_android.yml (hunk 15 above)
Comment (x3ro) [reply .github/workflows/ci_build_android.yml]: Done

**Comment by x3ro** [comment]
@acrrd @Robert-Steiner I think I've addressed everything, could you take another look?

**Comment by x3ro** [comment]
bors try

File: .github/actions/setup-job/action.yml [hunk 16]
```diff
@@ -50,6 +54,18 @@ runs:
         flutter-version: ${{ env.FLUTTER_VERSION }}
         cache: true
 
+    # - name: Install Android NDK
+    #   if: inputs.android != 'false'
+    #   uses: nttld/setup-ndk@4a541f506ff27d545f89e860b7cdc86efc3ca496 # HEAD 2022-02-15
+    #   with:
+    #     ndk-version: ${{ env.ANDROID_NDK_VERSION }}
```
Comment (Robert-Steiner) [inline .github/actions/setup-job/action.yml:61]:
```suggestion
```

File: .env [hunk 17]
```diff
@@ -11,6 +11,11 @@ FLUTTER_VERSION=2.5.3
 
 ANDROID_TARGETS="arm64-v8a x86_64 x86"
 ANDROID_PLATFORM_VERSION="21"
+# Note that this is _not_ the same as the platform version, and
+# does not need to align with it necessarily.
+# Versions can be found here https://developer.android.com/ndk/downloads
+ANDROID_NDK_VERSION="r23b"
```
Comment (Robert-Steiner) [inline .env:17]:
```suggestion
```

File: .github/workflows/ci_build_android.yml [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+        value: ${{ jobs.build-android-libs.outputs.artifact-dir-base }}
+
+jobs:
+  build-android-libs:
+    runs-on: ubuntu-20.04
+    timeout-minutes: 35
+    strategy:
+      matrix:
+        target: [
+          { rust: "aarch64-linux-android", ndk: "arm64-v8a" },
```
Comment (Robert-Steiner) [inline .github/workflows/ci_build_android.yml:17]: I'd rather just use the Rust target triple here to have less code

Comment (x3ro) [reply .github/workflows/ci_build_android.yml:17]: Since this mapping is not exactly obvious, I think it's good to have here in the CI. Especially considering that the rust triplets don't really have meaning within the Android build process, and we'd be naming the zip artifacts based on them. Following your suggestion, we would build a `build-android-libs-aarch64-linux-android.zip`, which would then contain a directory `arm64-v8a`, leaving the mapping to the person who has to deal with the artifact. 

Is this a must-have from your perspective?

Comment (Robert-Steiner) [reply .github/workflows/ci_build_android.yml:17]: > Following your suggestion, we would build a build-android-libs-aarch64-linux-android.zip, which would then contain a directory arm64-v8a, leaving the mapping to the person who has to deal with the artifact.

That's ok I think because these archives are just for sharing the artifacts between github jobs. It is not an artifact to be used by a person. The main purpose of the target name is to create a unique name for these artifacts within the matrix jobs

**Comment by x3ro** [comment]
bors try

File: .github/workflows/ci_build_android.yml (hunk 10 above)
Comment (Robert-Steiner) [reply .github/workflows/ci_build_android.yml:39]: I think the `CARGO_INSTALL_ROOT` is not a blocker for now. We can do it in a follow-up pr if we find a better solution. 
to make it a bit cleaner we could move this line to the step where we install cargo-ndk. 
but the line would need to change to: [`echo "${{ github.workspace }}/cargo-installs/bin" >> $GITHUB_PATH`](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#adding-a-system-path)
wdyt?

**Review by Robert-Steiner — APPROVED** [review]

File: .github/workflows/ci_build_android.yml (hunk 10 above)
Comment (rustonaut) [reply .github/workflows/ci_build_android.yml:39]: How is that failing, I mean sure `CARGO_INSTALL_ROOT` does install it somewhere else, but it also makes look `cargo` at that place when running a `cargo ..` sub-command :thinking: 

And `just` always loads `.env` files so the env should match.

Am I missing something?

**Comment by x3ro** [comment]
bors try

File: .github/workflows/ci_build_android.yml (hunk 10 above)
Comment (Robert-Steiner) [reply .github/workflows/ci_build_android.yml]: Seems that cargo ignores `CARGO_INSTALL_ROOT` when you try to run an executable :( They also point out adding the path:
```
  Installing cargo-installs/bin/async-bindgen-gen-dart
   Installed package `async-bindgen-gen-dart v0.1.0 (https://github.com/xaynetwork/xayn_async_bindgen.git#fe195d99)` (executable `async-bindgen-gen-dart`)
warning: be sure to add `cargo-installs/bin` to your PATH to be able to run the installed binaries
```

Comment (rustonaut) [reply .github/workflows/ci_build_android.yml]: Yes, but it's to run them without going through cargo I thought.

I mean what point is there in being able to change the install root if it then is ignored?

Comment (x3ro) [reply .github/workflows/ci_build_android.yml]: I think whatever the reasoning may be, `CARGO_INSTALL_ROOT` appears to not automagically also make the sub-command available to cargo, which is why we need to modify the `PATH` for `cargo-ndk` to work. This is necessary because we also can't invoke the `cargo-ndk` binary directly. I will go ahead and merge once the current state passes, then we can think about a different solution later on.

**Comment by x3ro** [comment]
bors merge
